File: scala/sdktest/PartnerEntityTest.scala
// Generated basic-flow test for the partner entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped PartnerTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefinshieldconexmgmtsdk.core.{Helpers, BluefinShieldconexMgmtSDK}
import voxgig.bluefinshieldconexmgmtsdk.utility.struct.Struct

object PartnerEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("partner.instance") {
      val testsdk = BluefinShieldconexMgmtSDK.testSDK()
      val ent = testsdk.partner(null)
      rep.check("partner.instance", ent != null, "expected non-null partner entity")
    }

    rep.scope("partner.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/partner/PartnerTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinShieldconexMgmtSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("partner01", "PARTNER01")
      idmap.put("partner02", "PARTNER02")
      idmap.put("partner03", "PARTNER03")
      val now = System.currentTimeMillis()

      // CREATE
      val partnerRef01Ent = client.partner(null)
      var partnerRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.partner"), "partner_ref01"))
      val partnerRef01DataResult = partnerRef01Ent.create(partnerRef01Data, null)
      partnerRef01Data = Helpers.toMapAny(partnerRef01DataResult)
      rep.check("partner.create.map", partnerRef01Data != null, "expected create result to be a map")
      rep.check("partner.create.id", partnerRef01Data != null && partnerRef01Data.get("id") != null, "expected created entity to have an id")

      // LIST
      val partnerRef01Match = new LinkedHashMap[String, Object]()
      val partnerRef01ListResult = partnerRef01Ent.list(partnerRef01Match, null)
      rep.check("partner.list.islist", partnerRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + partnerRef01ListResult)
      val partnerRef01List = partnerRef01ListResult.asInstanceOf[JList[Object]]
      val partnerRef01ListFound = Struct.select(
          SdkTestSupport.entityListToData(partnerRef01List), SdkTestSupport.om("id" -> partnerRef01Data.get("id")))
      rep.check("partner.list.exists", !Struct.isempty(partnerRef01ListFound), "expected to find created entity in list")

      // LOAD
      val partnerRef01MatchDt0 = new LinkedHashMap[String, Object]()
      partnerRef01MatchDt0.put("id", partnerRef01Data.get("id"))
      val partnerRef01DataDt0Loaded = partnerRef01Ent.load(partnerRef01MatchDt0, null)
      val partnerRef01DataDt0LoadResult = Helpers.toMapAny(partnerRef01DataDt0Loaded)
      rep.check("partner.load.map", partnerRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("partner.load.id", partnerRef01Data.get("id"), partnerRef01DataDt0LoadResult.get("id"))
    }
  }
}

File: rust/tests/partner_entity_test.rs
// Generated basic-flow test for the partner entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use bluefin_shieldconex_mgmt_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use bluefin_shieldconex_mgmt_sdk::utility::voxgigstruct as vs;
use bluefin_shieldconex_mgmt_sdk::{test_sdk, Entity, BluefinShieldconexMgmtEntity, BluefinShieldconexMgmtSDK, Value};

#[test]
fn partner_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.partner(Value::Noval);
    assert_eq!(ent.get_name(), "partner");
}

#[test]
fn partner_entity_stream() {
    // stream() runs the list op through the full pipeline and yields each
    // result item. Seed two entities via test mode; with the `streaming`
    // feature active it yields the feature's incremental items, else it
    // falls back to the materialised items — either way every item yields.
    let seed = jo(vec![(
        "entity",
        jo(vec![(
            "partner",
            jo(vec![
                ("strm01", jo(vec![("id", Value::str("strm01"))])),
                ("strm02", jo(vec![("id", Value::str("strm02"))])),
            ]),
        )]),
    )]);

    let sdkopts = jo(vec![(
        "feature",
        jo(vec![("streaming", jo(vec![("active", Value::Bool(true))]))]),
    )]);

    let testsdk = test_sdk(seed.clone(), sdkopts);
    let ent = testsdk.partner(Value::Noval);
    let items: Vec<Value> = ent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(items.len(), 2, "stream should yield both seeded items");

    // Fallback: streaming inactive still yields both materialised items.
    let plainsdk = test_sdk(seed, Value::Noval);
    let plainent = plainsdk.partner(Value::Noval);
    let plain_items: Vec<Value> = plainent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(plain_items.len(), 2, "fallback stream should yield both items");
}

#[test]
fn partner_entity_basic() {
    let setup = partner_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create", "list", "load"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("partner.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let partner_ref01_ent = client.partner(Value::Noval);
    let partner_ref01_data = to_map(&getp(
        &getpath(&["new", "partner"], &setup.data),
        "partner_ref01",
    ));

    let partner_ref01_data_result = partner_ref01_ent
        .create(partner_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let partner_ref01_data = to_map(&partner_ref01_data_result);
    assert!(
        matches!(partner_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );
    assert!(
        !getp(&partner_ref01_data, "id").is_noval(),
        "expected created entity to have an id"
    );

    // LIST
    let partner_ref01_match = Value::empty_map();

    let partner_ref01_list = partner_ref01_ent
        .list(partner_ref01_match.clone(), Value::Noval)
        .expect("list failed");
    assert!(
        matches!(partner_ref01_list, Value::List(_)),
        "expected list result to be an array"
    );

    let found_item = vs::select(
        &entity_list_to_data(&partner_ref01_list),
        &jo(vec![("id", getp(&partner_ref01_data, "id"))]),
    );
    assert!(
        !vs::is_empty(&found_item),
        "expected to find created entity in list"
    );

    // LOAD
    let partner_ref01_match_dt0 = jo(vec![("id", getp(&partner_ref01_data, "id"))]);
    let partner_ref01_data_dt0_loaded = partner_ref01_ent
        .load(partner_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    let partner_ref01_data_dt0_load_result = to_map(&partner_ref01_data_dt0_loaded);
    assert!(
        matches!(partner_ref01_data_dt0_load_result, Value::Map(_)),
        "expected load result to be a map"
    );
    assert_eq!(
        getp(&partner_ref01_data_dt0_load_result, "id"),
        getp(&partner_ref01_data, "id"),
        "expected load result id to match"
    );

}

fn partner_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("partner");
    entity_data_file.push("PartnerTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("partner01"), Value::str("partner02"), Value::str("partner03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID", idmap.clone()),
        ("BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value::str("FALSE")),
        ("BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value::str("FALSE")),
        ("BLUEFINSHIELDCONEXMGMT_APIKEY", Value::str("NONE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![("apikey", getp(&env, "BLUEFINSHIELDCONEXMGMT_APIKEY"))]), extra]),
            None,
        );
        BluefinShieldconexMgmtSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
